Added factory for RouteCollection, makes subclassing of RouteCollection possible.

lschricke opened the pull request:
Note: replaces #1174, which was targeting a wrong branch (and where I messed a bit with the commits ;)).

--

The aim of this PR is to allow developpers to subclass `Symfony\Component\Routing\RouteCollection` to change some of its behavior.

In most of my Silex apps, I figured I don't want the Symfony's `RouteCollection` regular behavior where if you define two routes with the same name (bad copy-paste for instance), the first route added simply gets overriden by the second. Instead, I would like to be able to throw an Exception if that happens. So I've written [a tiny subclass of RouteCollection](https://github.com/lschricke/symfony-strict-route-collection), but to be able to use it (through [a service provider like this](https://github.com/lschricke/silex-strict-route-collection-service-provider) for instance (needs to be updated for Silex 2)), I need to be able to replace `RouteCollection` by my subclass. This PR makes it possible.

Thanks! :)

fabpot commented on src/Silex/Application.php:
```diff
@@ -64,8 +63,12 @@ public function __construct(array $values = array())
 
-        $this['routes'] = function () {
-            return new RouteCollection();
+        $this['routes_class'] = 'Symfony\\Component\\Routing\\RouteCollection';
```
I would remove this as overriding the `routes_factory` would do the same. So, it's not really a good extension point.

fabpot commented on src/Silex/Application.php:
```diff
@@ -64,8 +63,12 @@ public function __construct(array $values = array())
-        $this['routes'] = function () {
-            return new RouteCollection();
+        $this['routes_class'] = 'Symfony\\Component\\Routing\\RouteCollection';
+        $this['routes_factory'] = $this->factory(function () use ($app) {
```
I don't see the need for this one as `$this['routes']` looks enough.

stof replied:
a factory is needed if ControllerCollection reuses it.

However @lschricke, this PR won't solve things for you if a ControllerCollection defines the same route name than another one (or that the main route collection)

lschricke replied:
Thanks for considering this PR and for the feedback! I'll look into it and update the PR soon.

fabpot commented on src/Silex/Application.php:
```diff
@@ -73,7 +76,7 @@ public function __construct(array $values = array())
 
         $this['controllers_factory'] = $this->factory(function () use ($app) {
-            return new ControllerCollection($app['route_factory']);
+            return new ControllerCollection($app, $app['route_factory']);
```
I would make the new argument second instead of prepending it to avoid an unneeded BC break. I would also inject the routes factory instead of the whole app.

fabpot commented on src/Silex/ControllerCollection.php:
```diff
@@ -38,13 +37,15 @@
     protected $prefix;
 
-    public function __construct(Route $defaultRoute)
+    public function __construct(Application $app, Route $defaultRoute)
```
So, the routes factory would be the second argument and it can even be optional. If not provided, just create a route collection like before.

fabpot commented on tests/Silex/Tests/ApplicationTest.php:
```diff
@@ -441,7 +441,7 @@ public function testRegisterShouldReturnSelf()
     {
         $app = new Application();
-        $mounted = new ControllerCollection(new Route());
+        $mounted = new ControllerCollection($app, new Route());
```
As the new argument is now optional, the tests can be keep like before, but new tests should be added to cover the new feature.

lschricke commented:
I updated the code and added some tests to cover this new feature as you suggested @fabpot. Do you think the tests provide enough coverage?

@stof, this PR does not solve my use case directly, but using it to have `Silex` use a subclass of Symfony's `RouteCollection` like [this one](https://github.com/lschricke/symfony-strict-route-collection) seems to work in my tests. Am I missing something?

fabpot commented on src/Silex/Application.php:
```diff
@@ -29,12 +29,12 @@
 use Silex\Api\ControllerProviderInterface;
 use Silex\Provider\RoutingServiceProvider;
 use Silex\Provider\KernelServiceProvider;
+use Symfony\Component\Routing\RouteCollection;
```
This should be reverted back to its original place.

fabpot commented on src/Silex/ControllerCollection.php:
```diff
@@ -11,8 +11,8 @@
 
-use Symfony\Component\Routing\RouteCollection;
 use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\RouteCollection;
```
should be moved back as before.

lschricke commented:
@fabpot PR updated.

lschricke commented:
I squashed my commits and rebase the PR.

fabpot commented:
Thank you @lschricke.

lschricke commented:
And thank you @fabpot and all the contributors for this awesome tool! :)
